### ISSUE TITLE
Code review - zuluzulu

### ISSUE THREAD

**Issue by dialexo** [njomeau/EPFLife#286]
This is a code review for the code of the whole project on master branch at commit 
https://github.com/njomeau/zuluzulu/tree/d327605a7a137853b10f5db0c8bcc65e22b26836

Of course, we don't expect you to change all of your code (some parts of the review can be subjective or arguable anyway), important is that you understand why better alternatives exist.

**Please DO ASK ME / US questions about anything that you do not understand, whether it is by writing a comment below or by sending us an email or meeting us.**

**Important : if you try to fix these, don't break your app. I prefer that the app works even if you don't manage to fix anything. It may be too hard to rework many of the issues below, as it would require to rewrite the whole architecture.**

## Packages

Your make good use of packages. Style note: usually package names start with lowercase.
A few small-ish remarks with the package hierarchy :

- Your `Structure` and `User` packages together kinda group all of the models in the MVC pattern. You could have a deeper hierarchy such as `models` with your Structure stuff inside and then `models/user` with your user classes.
- Same for your utility classes, you could have a single `utils` package with subpackages for `utils/url`, `utils/http`, `utils/test` for your IdlingResourceFactory, etc.
- I believe your packages for the database are reversed : it should be `Database/Firebase` rather than `Firebase/Database`. This actually shows the issues with your database abstraction, more on that later below.

## CommunicationTag.java

I'm not a huge fan of this enum with all the fragments tag. From my understanding, you mostly use it to define actions on the MainActivity. So you're basically delegating functionality to a big switch case : now imagine that we now want to delete some fragments and add some others: you then have to change your enum and your switch cases : this is not very maintainable and reliable. Also big red flag at this function signature : 

```java
public void onFragmentInteraction(CommunicationTag tag, Object data)
```
I don't think this is a good use case of `Object` : proof is that you're recasting data around afterwards.
Actually, why don't you approach this problem the other way around? A few insights :

- You already have nice factories in each of your fragments, so you can use dependency injection
- Your big onFragmentInteraction is basically delegating 3 functions of MainActivity (openFragment, selectItem, setTitle) so your code is already centralized
- When you use selectItem, you're basically delegating it again to another openFragment later on. Can't you call it directly?

So from my POV, your OnFragmentInteractionListener is adding indirection overhead without providing much of an abstraction (is it really a Listener anyway ? You're basically calling actions **on** it, so it doesn't look like one). I think a better implementation would be to pass a reference to the MainActivity in the factories of your fragment with DI. Then you can directly call your openFragment, setTitle and other MainActivity functionalities through this reference. In this way, you ensure that you get all the nice things from a typed setting (no cast of objects, and the compiler ensures that the functions exists and the types are correct). Of course, this would entail changing many implementation details in your current code so that we can call the functions from "the other side".

## ImageLoader

This is really nice, it really abstracts away the Glide implementation details, so that you could replace it with any other image loading library if required.

## Firebase

In the contrary, your database abstraction is very convoluted and would make it very difficult to swap with another database implementation.

First, it is a very good thing that you have tried to abstract the database implementation away from the rest of the codebase. Unfortunately, there are many issues with your current architecture, so please let me point out the flaws.

You have basically put every structural design pattern in your database code. However, more DP doesn't mean better code. It took me a while (1/2 hour) to figure out your class hierarchy.

So let me recap your current code structure : your entry point is DatabaseFactory, which retrieves a singleton instance of FirebaseProxy inheriting from a class named Proxy which holds many abstractions to interact with the database. Your FirebaseProxy then uses a FirebaseFactory to retrieve an instance of FirebaseAdapter which inherits from an interface named Database holding a single query function called collection. Your FirebaseAdapter holds an instance of the proprietary FirebaseFirestore instance, and its concrete definition of collection fetches a DatabaseCollection (which abstractly defines document to get a DatabaseDocument), through a CollectionAdapter. The CollectionAdapter holds CollectionReferences that are part of the proprietary Firebase implementation to your real data. The DatabaseDocument and its functionalities (noSQL-like operations such as set, update, get and collection to get child DatabaseDocuments) are the main API that is exposed to the rest of the app when the data is fetched. All these operations are reflected (delegated) through the DocumentAdapters wrapping all DatabaseDocuments, and they forward your calls directly to the proprietary Firebase API through the CollectionReferences.

I think that this is very complex an architecture, which unfortunately doesn't help much : it doesn't provide a clear API to your internal codebase, makes the code very hard to understand to another developer and also prevents a quick migration to another database technology. Let me explain every point :

### Your DatabaseFactory retrieves an external singleton

What is DatabaseFactory used for? It basically gets the singleton instance from FirebaseProxy : however the whole point of a singleton is to be self-contained and self-controlled. This extra indirection does not provide any abstraction for it. I understand that setDependency is used for testing and that it is where you swap it : this is a good example of where a singleton looks perfect in theory but falls apart in practice, as testing / mocking them without breaking their encapsulation is very hard (I do not blame you for that). An ideal solution would be using dependency injection to pass around your database (this is applicable even if it is a singleton fetch at instantiation), so that mocking it becomes easy (as you can swap the dependency at runtime directly in your fragments). As a bonus, you can express your fragment controller as a "function" of inputs, including the database, which produces a view. But unfortunately, the Android environment isn't too helpful as there are many other initialization steps in the fragments lifecycle. Still, since you can create them using factories, DI could be applicable. (Don't change this in your project now, this would require far too much work.)

A factory should allow an object to express how it want its instances to be created. Factories are nice patterns and are used often, but do not abuse them, they should be used when additional instantiation logic is required.

You have the same kind of indirection overhead with FirebaseFactory and FirebaseAdapter.

### You abstract the database the wrong way

This is the main problem of your whole implementation in my opinion. Basically, since you use low-level storage concerns as your top-level interface to communicate with the data and to interact with the database, your code cannot provide high-level semantics that would allow you to completely forget about the underlying database technology from the user side. Let me explain.

You use classes from the Firebase API (such as DocumentReference and CollectionReference) that you rewrite yourself (DatabaseDocument and DatabaseCollection) **at the same low-level of abstraction** and make them your top level definitions, and then use an Adapter to "convert" them (but in fact they don't adapt anything, you simply forward the same calls to the real Firebase collection expressed in their API). What I mean by *low-level of abstraction* is that your DatabaseDocument class exposes how to **store and read** the class as if it was a **file** that you store on the server, and this is what you expose as your internal API. By doing so, you lose all benefits from the typed way NoSQL works : this technology should help you talk in terms of objects only, and storage should not be a concern of the programmer. A better abstraction would be to have functions that effectively fetch the documents but **convert them at the same time** into your models paradigm. We want to be able to simply call `database.getAssociations()` and you retrieve a `List<Association>` for instance, or `database.getUser(id)` and you get a single `User` (btw these can be sync or async). I see that this is kind of what your "Proxy" does, but in fact it is what the core abstraction of the database should have been.

Bottom line : a good DB implementation is one that completely abstracts the underlying details of the technology and allows you to express your data in terms of regular data classes and their collections anywhere you use it (every call to `database`).

### Your DB is modeled after NoSQL

Another corollary of the point above is that migrating your current DB would be very hard. How many files, interfaces, implementation details would you need to modify if suddenly your product owner decided to switch to another database model, such as relational databases or graph-oriented databases ? Probably all of your database classes. With a good database abstraction, you would only need to :

1. Write a **single new** class implementing the high-level operations defined in the core abstraction with the implementation details of the specific technology
2. Modify **a single line of code** where your database gets instantiated. Then, thanks to dependency injection, everything works out of the box.

### Your proxy is not a proxy

The abstract functions you've put in Proxy (and its definitions in FirebaseProxy) are actually very good, but they are not at the right place and do not serve the purpose of a Proxy pattern.
Actually, you are **not implementing a Proxy DP**, and **you don't need one**. I suspect the "Proxy" is one of the last part of the DB codebase you have written as it kind of "fixes" everything else.
As mentionned above, these function definitions should have been in your top-level database interface : they express everything we want to do with the database in model-land, and the concrete child takes care of the proprietary Firebase codebase. And it should not be named Proxy. Simply Database would fit this class better.

Proxy is a pattern where you share the exact same interface with another object so that you can swap them indistinguishably. This is useful for instance when you want additional control in-between the desired class and the call site because you can wrap additional logic before forwarding the calls to a real object. A database proxy could be a valid use-case, for instance if we want to implement redundant storage (one can imagine that if one database is detected as down, then the proxy could forward to another live db). It could also log / monitor traffic or ensure that the data is always valid.

Another issue is that when you fetch your database in your fragments, the type associated to its variable is `Proxy`. This makes it very obscure for another programmer to understand that it is in fact the database. Your top-level type should be something similar to `Database`. Also, if it was a real Proxy, since it shares its inherited interface with the object it should replace, you should not be able to see that it is indeed one from the type alone.

### Suggestions for a better implementation

- Use abstractions : your pseudo-"Proxy" does that nicely. Build them bottom up and not top down and use them at the highest level in the type hierarchy, so that you can express complex operations in terms of ideas and models.
- Don't use indirection abusively : remember that each forwarding is also a performance loss
- Simplify your class hierarchy : abstractions are important, but they must remain meaningful.
- Use dependency injection : especially since you have Fragments with factories. This will allow you to declare the instantiation of dependences only once, and then cascade the references
- The overall design should stay simple : we want it to be hard to break, and easy to work with.
- DPs are good but know when and why you use them.
- Think of swappability : minimize the amount of code that would need to be changed if a dependency needs to be swapped.
- Class naming is important. Some of your class names do not accurately describe what they are in fact really implementing. Imagine that a new programmer should be able to instinctively understand your code.

## Models

Your model classes in package Structure are quite good (nice use of the Builder pattern). Many of your classes are read-only (with constructors and getters only), is there a reason why they are not immutable? Remember that eliminating state can remove many classes of nasty bugs. Remember to add the `final` keyword, to deep copy and freeze referenced properties when you construct them and also when they are retrieved with a getter (such as lists and objects). Also, since many of these are data classes, it may be useful to re-implement `equals`, `clone` and the like, so that you can treat theses classes the same as values.

## Android stuff

**Hardcoded strings / values** : you have some hardcoded strings in your code and also in your layout XML files. This makes it difficult to maintain and translate, Android Studio can help you transform them automatically into templates (using strings.xml).

**Be consistent with your code comments** : some of your classes are very well documented while others are devoid of them. Develop good habits and write Javadoc as you code. It helps your colleagues and your future self.

**Tests** : Sometime you repeat some code or call a setup function before / after every test. A function call is fine, but fyi JUnit has dedicated annotations to setup the test class, setup before every test, etc.. (`@Before`, `@BeforeClass`, etc. have a look at the docs)

There's a stray character in your AndroidManifest.xml at line 11.

**Comment by Dahnepfl** [comment]
Hello @dialexo 

Thank you for your great review !

Can we meet on wednesday at 10am (in exercice session ?) to discuss about this issues ? 

Thank you

Dahn

**Comment by Dahnepfl** [comment]
I thought that we had an exercice session instead.

Fine, see you on Wednesday :)

**Comment by Dahnepfl** [comment]
Will be here at 10am

**Comment by dialexo** [comment]
Hey guys I just got sick today, I'll see if I'm better tomorrow. If not, would videoconference be okay? 
@dahnepfl @betty

**Comment by Dahnepfl** [comment]
Oh, I hope you will soon recover !
I think that I am the only guy of the group to come. I just wanted to speak about our mistakes and how we could have avoided them.

I am not very confortable with a videconference. Is it possible to meet an other day ? If not, If @betty is here to help me set the videoconference, I will be there at 10 am.

**Comment by dialexo** [comment]
Thanks Dahn, let's reschedule this then. What about Friday, between 3pm and 6pm?

**Comment by Ikaras998** [comment]
Oh if it is friday around this time i will be available to come as well

**Comment by dialexo** [comment]
Hello @Dahnepfl, @Ikaras998, it looks like I've been too optimistic about the recovery time as I am unfortunately still sick. I apologize for the inconvenience and can only propose to reschedule it early next week. Assuming I'll be fine by then would Monday between 2pm and 7pm work?

**Comment by Dahnepfl** [comment]
Yeah it works for me :)

Hope you will soon get better

**Comment by Ikaras998** [comment]
Hello @dialexo
I apologize but I have something unplanned that I will have to attend this afternoon, but it s not a problem if you have the meeting without me today, even more that I got the answers to my questions already by chance. But thank you and I hope you feel better today

**Comment by Dahnepfl** [comment]
I will be here at 5pm for a short period of time to ask few questions (I have an exam tomorow).

**Comment by dialexo** [comment]
Yeah it's fine you will find me in the usual room at 5

**Comment by Dahnepfl** [comment]
I m on my way I arrive in 5mn sorry
